runtime/src/accounts_index.rs

```diff
@@ -32,7 +32,164 @@ pub const ITER_BATCH_SIZE: usize = 1000;
 
 pub type SlotList<T> = Vec<(Slot, T)>;
 pub type SlotSlice<'s, T> = &'s [(Slot, T)];
-pub type Ancestors = HashMap<Slot, usize>;
+
+pub type AncestorsForSerialization = HashMap<Slot, usize>;
+
+#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize, AbiExample)]
+pub struct Ancestors {
+    min: Slot,
+    slots: Vec<Option<usize>>,
+    count: usize,
+    max: Slot,
+    large_range_slots: HashMap<Slot, usize>,
+}
+
+// some tests produce ranges that are too large such
+// that we don't want to represent them in the optimal way.
+// so fallback to HashMap
+const ANCESTORS_HASH_MAP_SIZE: u64 = 10_000_000;
+
+impl From<Vec<(Slot, usize)>> for Ancestors {
+    fn from(source: Vec<(Slot, usize)>) -> Ancestors {
+        let mut result = Self {
+            count: source.len(),
+            ..Ancestors::default()
+        };
+        if result.count > 0 {
+            result.min = Slot::MAX;
+            result.max = Slot::MIN;
+            source.iter().for_each(|(slot, _)| {
+                result.min = std::cmp::min(result.min, *slot);
+                result.max = std::cmp::max(result.max, *slot + 1);
+            });
+            let range = result.range();
+            if range > ANCESTORS_HASH_MAP_SIZE {
+                result.large_range_slots = source.into_iter().collect();
+            } else {
+                result.slots = vec![None; range as usize];
+                source.into_iter().for_each(|(mut slot, size)| {
+                    slot -= result.min;
+                    result.slots[slot as usize] = Some(size);
+                });
+            }
+        }
+
+        result
+    }
+}
+
+impl From<&HashMap<Slot, usize>> for Ancestors {
```

Comment:
vec and hashmap are very similar code. I spent some time with iterators, generics, traits, etc. trying to get a cleaner solution than duplicating the code. So far, my efforts have not worked.

Reply:
hashmap is needed for persistence, vector is the way all production code initializes a vector (as well as most test code).